stan/math/rev/core/chainable_object.hpp

```diff
@@ -0,0 +1,35 @@
+#ifndef STAN_MATH_REV_CORE_CHAINABLE_OBJECT_HPP
+#define STAN_MATH_REV_CORE_CHAINABLE_OBJECT_HPP
+
+#include <stan/math/rev/meta.hpp>
+#include <stan/math/rev/core/chainable_alloc.hpp>
+#include <stan/math/rev/core/typedefs.hpp>
+#include <stan/math/prim/fun/Eigen.hpp>
+#include <stan/math/prim/fun/typedefs.hpp>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+template <typename T>
+class chainable_object : public chainable_alloc {
+private:
+  T obj_;
+
+public:
+  template <typename S>
+  chainable_object(S&& obj) : obj_(std::forward<S>(obj)) {}
+
+  T& get() { return obj_; }
+};
+
+template <typename T>
+auto make_chainable_ptr(T&& obj) {
+  auto ptr = new chainable_object<plain_type_t<T>>(std::forward<T>(obj));
+  return ptr;
+}
+
```

Comment:
@t4c1 @SteveBronder 

The goal here is to create ptrs to Eigen objects in the forward pass that will get deallocated with `stan::math::recover_memory` and `stan::math::recover_memory_nested`.

I made a template godbolt with this to see if it was forwarding stuff along but it has a lot of code in it and is confusing (presumably from the QR). Maybe it is useful to you all: https://godbolt.org/z/e1eMof

Reply:
Questions are:

1. Does this pattern look okay?
2. Am I duplicating something that exists elsewhere?

If the answers are yes/no, then I'll go add tests and comments and whatnot.

Reply:
Talked to Steve. Tests added.

Reply:
That is a nice idea. I like it. However I am not sure if we always want to store only plain types. I guess that should be up to the user of this?

Reply:
> I am not sure if we always want to store only plain types. I guess that should be up to the user of this?

How about I replace `plain_type_t` with `std::decay`?

Reply:
That should be fine.

Reply:
@t4c1 I was using this and I realized that maybe I should just return a pointer to the underlying object.

Like instead of:

```
return new chainable_object<plain_type_t<T>>(std::forward<T>(obj));
```

```
return &(new chainable_object<plain_type_t<T>>(std::forward<T>(obj)))->get();
```

So that way nobody has to deal with `chainable_objects` directly.

Reply:
Or what if this was the default implementation of `to_arena`?

```
return (new chainable_object<plain_type_t<T>>(std::forward<T>(obj)))->get();
```

So then if something can't be shipped off to the arena, `to_arena` has the effect of giving it arena memory lifetime.

Reply:
Wait I guess the thing about `arena_t` is that it gives you lightweight variables that you are cheap to copy.

A reference to the underlying object is not cheap or safe to copy.

Reply:
I changed this to return a pointer to the underlying object (instead of the `chainable_object` itself)